File: language_construct_performance/python/scripts/value_obj_minimal_variable_properties.py
import os
import gc
from dataclasses import dataclass
from typing import List, Optional, Any, Dict
from benchmark_utils import NUM_ENTRIES, benchmark_stats, save_stats

# --- VALUE OBJECT CLASSES ---

@dataclass(slots=True)
class MinimalValueObjectNode:
    label: str
    id: str
    accessTo: List[str]
    details: Dict[str, Any]
    edgeTo: Optional[List[str]] = None

# --- FACTORIES ---

def create_value_object(index: int) -> MinimalValueObjectNode:
    edgeTo = ['r.392', 'r.40', 'r.41', 'update', 'administrator', 'create', 'delete', 'read'] if index % 2 == 0 else None
    ut = 2 if index % 3 == 0 else None
    ag = None
    if index % 4 == 0:
        ag = {
            's': {'t': 167},
            'a': {
                't': 3187978,
                's': {
                    't': 3187978,
                    's': 3149311,
                    'r': 42506
                }
            }
        }

    details = {
        'provider': 'aws',
        'accountId': '568709751681',
        'principal': True,
        'tags': ['aKIAYI2NaRQPOT', 'dev testing local'],
        'mfas': '',
        'la': 1772454942 + (index % 1000),
        's': 1,
        'cpd': 0,
        'pcb': '-',
        'lld': 0,
        'cd': 1763097939000,
        'cb': '-',
        'ub': '-',
        'ud': 0,
        'ua': 1772526871591
    }

    if ut is not None:
        details['ut'] = ut
    if ag is not None:
        details['ag'] = ag

    return MinimalValueObjectNode(
        label=f'user-dev-test-{index}',
        id=f'u.{16406 + index}',
        accessTo=['s.[s3].UACDR', 'a.[s3].DARC', 's.[secretsmanager].RACDU', 's.[dynamodb].RCDAU'],
        details=details,
        edgeTo=edgeTo
    )

# --- BENCHMARK RUNNER ---

def run_benchmark():
    stats = {}
    print(f"Starting Value Object Variable Properties Minimal Benchmark with {NUM_ENTRIES:,} entries...\n")

    value_obj_list = [None] * NUM_ENTRIES

    print("\n--- CREATION ---")
    def creation():
        for i in range(NUM_ENTRIES):
            value_obj_list[i] = create_value_object(i)
    benchmark_stats("Value Object Minimal Creation", stats, "creationTimeMs", creation, track_memory=True)

    print("\n--- PLAIN TRAVERSAL ---")
    def traversal():
        dummy_count = 0
        for i in range(NUM_ENTRIES):
            if value_obj_list[i] is not None:
                dummy_count += 1
        return dummy_count
    benchmark_stats("Plain Traversal (Value Object Minimal)", stats, "plainTraversalTimeMs", traversal)

    print("\n--- PROPERTY ACCESS ---")
    def property_access():
        sum_val = 0
        for i in range(NUM_ENTRIES):
            ag = value_obj_list[i].details.get('ag')
            if ag:
                sum_val += ag['a']['s']['r']
        return sum_val
    benchmark_stats("Property Access (Value Object Minimal)", stats, "propAccessTimeMs", property_access)

    print("\n--- FILTERING ---")
    def filtering():
        matched = 0
        for i in range(NUM_ENTRIES):
            if value_obj_list[i].details['la'] > 1772455500:
                matched += 1
        return matched
    benchmark_stats("Filtering (Value Object Minimal)", stats, "filterTimeMs", filtering)

    print("\n--- MUTATION ---")
    def mutation():
        for i in range(NUM_ENTRIES):
            value_obj_list[i].details['la'] += 1
    benchmark_stats("Mutation (Value Object Minimal)", stats, "mutationTimeMs", mutation)

    print("\n--- DELETE PROPERTY ---")
    def delete_property():
        for i in range(NUM_ENTRIES):
            if 'ud' in value_obj_list[i].details:
                del value_obj_list[i].details['ud']
    benchmark_stats("Delete Property (Value Object Minimal)", stats, "deletePropertyTimeMs", delete_property)

    # Clear memory
    value_obj_list.clear()
    gc.collect()

    # Save Stats
    save_stats("stats_variable.json", "value object minimal", stats)

if __name__ == "__main__":
    try:
        run_benchmark()
    except Exception as e:
        print(f"Error: {e}")
        import sys
        sys.exit(1)

File: language_construct_performance/ruby/scripts/value_obj_minimal_variable_properties.rb
require_relative 'benchmark_utils'

# --- VALUE OBJECT CLASSES ---
# Minimal: only the top-level node is a class, details remain a plain Hash

class MinimalValueObjectNode
  attr_accessor :label, :id, :access_to, :details, :edge_to

  def initialize(label, id, access_to, details, edge_to: nil)
    @label = label
    @id = id
    @access_to = access_to
    @details = details
    @edge_to = edge_to if edge_to
  end
end

# --- FACTORIES ---

def create_value_object(index)
  edge_to = index % 2 == 0 ? ["r.392", "r.40", "r.41", "update", "administrator", "create", "delete", "read"] : nil
  ut = index % 3 == 0 ? 2 : nil
  ag = index % 4 == 0 ? {
    s: { t: 167 },
    a: { t: 3187978, s: { t: 3187978, s: 3149311, r: 42506 } }
  } : nil

  details = {
    provider: "aws",
    account_id: "568709751681",
    principal: true,
    tags: ["aKIAYI2NaRQPOT", "dev testing local"],
    mfas: "",
    la: 1772454942 + (index % 1000),
    s: 1,
    cpd: 0,
    pcb: "-",
    lld: 0,
    cd: 1763097939000,
    cb: "-",
    ub: "-",
    ud: 0,
    ua: 1772526871591
  }

  details[:ut] = ut if ut
  details[:ag] = ag if ag

  MinimalValueObjectNode.new(
    "user-dev-test-#{index}",
    "u.#{16406 + index}",
    ["s.[s3].UACDR", "a.[s3].DARC", "s.[secretsmanager].RACDU", "s.[dynamodb].RCDAU"],
    details,
    edge_to: edge_to
  )
end

# --- BENCHMARK RUNNER ---

stats = {}

puts "Starting Value Object Variable Properties Minimal Benchmark with #{NUM_ENTRIES.to_s.gsub(/(\d)(?=(\d{3})+$)/, '\\1,')} entries..."

value_obj_array = Array.new(NUM_ENTRIES)

puts "\n--- CREATION ---"
benchmark_stats("Value Object Minimal Creation", stats, "creationTimeMs", track_memory: true) do
  NUM_ENTRIES.times do |i|
    value_obj_array[i] = create_value_object(i)
  end
end

puts "\n--- PLAIN TRAVERSAL ---"
benchmark_stats("Plain Traversal (Value Object Minimal)", stats, "plainTraversalTimeMs") do
  dummy_count = 0
  NUM_ENTRIES.times do |i|
    dummy_count += 1 if value_obj_array[i]
  end
  dummy_count
end

puts "\n--- PROPERTY ACCESS ---"
benchmark_stats("Property Access (Value Object Minimal)", stats, "propAccessTimeMs") do
  sum = 0
  NUM_ENTRIES.times do |i|
    ag = value_obj_array[i].details[:ag]
    sum += (ag ? ag[:a][:s][:r] : 0)
  end
  sum
end

puts "\n--- FILTERING ---"
benchmark_stats("Filtering (Value Object Minimal)", stats, "filterTimeMs") do
  matched = 0
  NUM_ENTRIES.times do |i|
    matched += 1 if value_obj_array[i].details[:la] > 1772455500
  end
  matched
end

puts "\n--- MUTATION ---"
benchmark_stats("Mutation (Value Object Minimal)", stats, "mutationTimeMs") do
  NUM_ENTRIES.times do |i|
    value_obj_array[i].details[:la] += 1
  end
end

puts "\n--- DELETE PROPERTY ---"
benchmark_stats("Delete Property (Value Object Minimal)", stats, "deletePropertyTimeMs") do
  NUM_ENTRIES.times do |i|
    value_obj_array[i].details.delete(:ud)
  end
end

# Clear memory
value_obj_array.clear
value_obj_array = nil
GC.start

# Save Stats
save_stats("stats_variable.json", "value object minimal", stats)
